File: python/algorithms/merge sort.py
def merge_sort(arr):
    if len(arr)>1:
        mid = len(arr)//2
        l = arr[:mid]
        r = arr[mid:]
        merge_sort(l)
        merge_sort(r)
        i,j,k = 0,0,0
        while i < len(l) and j<len(r):
            if l[i]<r[j]:
                arr[k] = l[i]
                i+=1
            else:
                arr[k] = r[j]
                j+=1
            k+=1
        while i<len(l):
            arr[k] = l[i]
            k+=1
            i+=1
        while j<len(r):
            arr[k] = r[j]
            k+=1
            j+=1
arr = [5,2,4,6,1,3]
merge_sort(arr)
print(arr)

File: haskell/algorithms/mergeSort.hs
merge :: Ord a => [a] -> [a] -> [a]
merge [] xs = xs
merge xs [] = xs
merge (x1 : xs1) (x2 : xs2)
  | x1 < x2 = x1 : merge xs1 (x2 : xs2)
  | otherwise = x2 : merge xs2 (x1 : xs1)

mergeSort :: Ord a => [a] -> [a]
mergeSort [] = []
mergeSort [x] = [x]
mergeSort xs = merge l r
  where
    lr = splitAt (length xs `div` 2) xs
    l = mergeSort $ fst lr
    r = mergeSort $ snd lr

main :: IO ()
main = do
  let a = [1,2,0,5]
      b = mergeSort a
  print b
